### DISCUSSION TITLE
New OAuth2 login flow is not supported on installed PWAs on iOS

### DISCUSSION THREAD

**Discussion by Fischer96** [pocketbase/pocketbase#2429] [Q&A] [answered]
Hey, i just moved to the new login flow and the popup was blocked in desktop. As it can be allowed here easily everything was fine so far but on iOS this popup will never show up when the web app is installed as PWA on the home screen. Is there any way around that problem?

**Comment by ganigeorgiev** [comment]
I don't have an iOS device to test it, but you could try to specify your own url opener action using the `urlCallback` prop:

```js
await pb.collection("users").authWithOAuth2({
    provider:    "google",
    urlCallback: (url) => {
        // open the url by some js method that is allowed
        window.open(url, "_blank", "");
    },
})
```

If you can't find a js action that can open safely the above url, you can try to either setup a proxy redirect (eg. `'yourdomain.com?redirect=' + url` and then redirect to the `url` query param value; _I'll consider in the future to see how feasible that would be to have it builtin_) OR use the manual OAuth2 code exchange flow.

**Reply by ganigeorgiev** [reply]
Reading https://web.dev/learn/pwa/windows/#forcing-a-browsers-navigation, it may not be possible to open a new browser window from PWA on mobile device:

> If you want to force open the browser with a URL and not an in-app browser, you can use the _blank target of `<a href>` elements. This works only on desktop PWAs; **on mobile devices, there is no option to open a browser with a URL**.

**Reply by ganigeorgiev** [reply]
Further reading the post from the above link:

> Your PWA redirects the user to a **URL that is out of the scope of the PWA** so that the rendering engine **will open an in-app browser within your PWA**.

Or in other words, you can try in the above `urlCallback` instead of explicitly opening a new window to redirect to the url, aka. `window.location.href = url`.

If that doesn't work, then I guess the manual OAuth2 code exchange flow remains an option since you'll have full control over the redirect flow.

**Comment by Fischer96** [comment]
I already tried the `window.location.href` trick but that results in this error message from pocketbase:

```
{
  "code": 404,
  "message": "Missing or invalid OAuth2 subscription client.",
  "data": {
    
  }
}
```

**Reply by Fischer96** [reply]
The popup was blocked because i automatically called this auth process when the user is not logged in. Apparently browsers only allow popups when this auth process is triggered manually via a button for example. I guess for now it would also work to manually press the login button

**Reply by ganigeorgiev** [reply]
When you call `window.location.href` does it redirect, aka. replaces the current pwa page with the new one or opens a new in-app browser browser? Maybe the behavior of the redirect differ on desktop and mobile?

**Reply by Fischer96** [reply]
It will redirect to the spotify login page in my case. Once i press "login" on spotify it runs into that error. On phones it will open a new in app browser fine but yeah, same behavior, it will run into that error page

**Reply by ganigeorgiev** [reply]
>  On phones it will open a new in app browser fine but yeah, same behavior, it will run into that error page

If the original page where you've called `authWithOAuth2` is not closed, aka. the connection is still open, then the 404 error shouldn't occur.

**Reply by Fischer96** [reply]
Yeah there seems to be no way without doing it manually via a button press. Sad that apple doesn't allow it. Thank you anyways!

**Reply by ganigeorgiev** [reply]
_Just to clarify since I'm not sure what do you refer by "doing it manually via a button press"._

Yes, you need the user to manually open the url, aka. to show a "Login with ..." btn that will call `authWithOauth2`, but if `window.location.href` in a PWA on mobile device opens **new** in-app browser, then the OAuth2 flow should work fine and you'll not get the above 404 error. You could add a condition in the `urlCallback` that will use `window.open` for desktop and `window.location.href` on mobile, something like:

```js
await pb.collection("users").authWithOAuth2({
    provider:    "google",
    urlCallback: (url) => {
        // detect mobile device based on useragent or some other special navigator.* field
        const isMobile = ...

        if (isMobile) {
            window.location.href = url;            
        } else {
            window.open(url, "_blank", "");
        }
    },
})
```

**Reply by philffm** [reply]
Cheers, I just stumbled on the issue as well and came to try your proposed approach.

The twitter page opens but after authentication, it redirects to here:

![image](https://user-images.githubusercontent.com/6079545/236892044-4cb09576-7e2d-40d0-9e6d-2984a074163f.png)

https://server.com/api/oauth2-redirect?state=AJPUVAPAXXXXXXXXXXXXXXXXXXXO9BiAsTd&code=NFU1b1lyXXXXXXXXXXXXXXXXXXOjE

```

const authData = await this.$pb.collection("users").authWithOAuth2({
            provider:    "twitter",
                urlCallback: (url) => {
                    // detect mobile device based on useragent or some other special navigator.* field
                    const isMobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);

                    if (isMobile) {
                        window.location.href = url;            
                      } else {
                      window.open(url, "_blank", "");
                    }
                },
            })
```

**Reply by ganigeorgiev** [reply]
@philffm Are you sure that the regex works? The above will work only if:

- `location.href` actually opens an in-app browser in the PWA
- the PWA page that initialized the connection is not closed and is running in the background

If the above doesn't work, then you'll have to implement the manual code exchange flow. I don't know if there is other workaround.

**Reply by ganigeorgiev** [reply]
There is also the [`scope` manifest property](https://web.dev/add-manifest/#scope) that eventually could help, but I haven't tested it:

> If the user clicks a link in your app that navigates outside of the scope, the link opens and renders within the existing PWA window. **If you want the link to open in a browser tab, you must add target="_blank" to the <a> tag. On Android, links with target="_blank" open in a [Chrome Custom Tab](https://developer.chrome.com/multidevice/android/customtabs).**

Or in other words, you can try to specify the scope property so that external links can open in a new window/tab outside of the PWA. Then you can create a dummy link programmatically to trigger the url, something like:

```js
const tempLink = document.createElement("a");
tempLink.setAttribute("href", url);
tempLink.setAttribute("target", "_blank");
tempLink.click();
tempLink.remove();
```

**Comment by driouxg** [comment, marked as answer]
Using @philffm solution and this answer from [Stack Overflow](https://stackoverflow.com/questions/20696041/window-openurl-blank-not-working-on-imac-safari) I've got something that's working for me to be able to oauth on iOS without having the popup being blocked and without having to prompt the user to turn off their popup blocker in iOS. Basically, iOS won't open new windows inside of async functions, so open the window before calling the .authWithOAuth2() function.

```javascript
let w = window.open()

pocketbaseClient
    .collection("users")
    .authWithOAuth2({
      provider: provider.name,
      urlCallback: (url) => {
          w.location.href = url
      },
    })
```

**Reply by philffm** [reply]
Happy to hear it works for you 🙌 Did it the manual way in the meantime which also works great but might try your code as well

**Reply by hungcrush** [reply]
@driouxg Thank you very very much....

**Reply by jeffrey0606** [reply]
Thank you @driouxg works perfectly

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here - with JS SDK v0.17.0+ the SDK opens eagerly the window before the async call (I don't have Mac or Safari to verify it). **Note that you still may have to manually open eagerly the window on your own if the `authWuthOauth2` call is executed within another wrapped async function.**_

**Reply by jlak0** [reply]
all work fine, but why this method does not create new users?
authentication fail if user does not exist

**Reply by ganigeorgiev** [reply]
@korosu I don't think your issue is related to the above discussion.

Double check your users collection create API rule and fields constraints (eg. required or unique fields). Try to inspect the returned error response by wrapping the call in a `try/catch` and add `console.log(err.response)`. Alternatively, you can also check the failed request logs in the _Admin UI > Logs_ for more details.

If you are not able to resolve it, please open a new discussion with more details about your use case, the error that is returned and some steps to reproduce.

**Reply by jlak0** [reply]
Thank you for your help.
I checked the logs and found the issue is related to API rules. The first time, was I had an admin account logged in, so I created a new user without issue.
Now, I have set the create permission to public, and everything works perfectly.

**Reply by Brenden-SSI** [reply]
This worked for me too, I had the window opening in an async function, moved it out to a new one : 

```
function openAuthWindow() {
    authWindow = window.open('about:blank', '_blank');
}
```

and then have the button fire both (svelte)

```
<button class="google-button" on:click={() => { openAuthWindow(); handleGoogleLogin(); }}>
    <img src="https://www.gstatic.com/firebasejs/ui/2.0.0/images/auth/google.svg" alt="Google logo" />
    Sign in with Google
</button>

```

**Reply by dahjson** [reply]
This method still doesn't work in Safari. Is there any way to just stay in the same window and not open another tab or new popup window?

**Reply by dahjson** [reply]
This should be kept in the same window. No popups or new tabs. This can all get blocked by browsers like Safari and ad blocker extensions. Being able to do something more like this would be ideal. It should be the default as well, so the callback isn't even required.

```
pocketbaseClient
  .collection("users")
  .authWithOAuth2({
    provider: provider.name,
    urlCallback: (url) => {
      window.location.href = url
    },
    redirectUrl: https://example.com
  })
```

**Reply by ganigeorgiev** [reply]
@dahjson Please stop spamming about the same thing in multiple places. As mentioned in your previous opened issues there are no plans to change the current implementation.

As confirmed by many other users, the Safari issue can be resolved by simply triggering the auth call in a **non-async** handler. You can also check for a live example in https://app.presentator.io/#/login (the related code could be found in https://github.com/presentator/presentator/blob/09609c49c67275bf42ec7769d59a9d3d72f63645/ui/src/components/base/SocialAuth.svelte#L9).

If you want to have full control over the redirect handler then you can implement the "Manual code exchange" flow (the second tab in https://pocketbase.io/docs/authentication/#oauth2-integration).

**Comment by tschmidt-code** [comment]


**Reply by ganigeorgiev** [reply]
You are getting a blank page because on click you are opening a new window but the `authWithOAuth2` call probably throw an error because it cannot connect to your PocketBase server (try to replace the console.log with window.alert to confirm it if you can't access the inspect tool).

`localhost` in the above need to be replaced with your public address because the mobile device has its own localhost and it cannot connect to your PocketBase server.

Also note that opening manually a window outside of the `authWithOAuth2` call is not required if you use the latest JS SDK.

**Reply by ganigeorgiev** [reply]
For testing on mobile I'd recommend to use ngrok, zrok or other ssh tunneling service that will create a public https address for your localhost.

Note also that the redirect url depends on the url you use in the JS SDK initialization, aka. if you use the IP in `new PocketBase(http://IP:8090)`, then the callback url that you have to setup in the provider's OAuth2 app must be something like: `http://IP:8090/api/oauth2-redirect`.

**Comment by sixthkind** [comment]
I stumbled upon this thread when looking for a solution to a capacitor app using pocketbase.

You can use @capacitor/browser as a solution.
https://ionicframework.com/docs/native/browser

Instead of using `window.location.href = url` in the `urlCallback`, use the in app browser with `Browser.open({ url })`:

```
import { Browser } from '@capacitor/browser';

pocketbaseClient
    .collection("users")
    .authWithOAuth2({
      provider: provider.name,
      urlCallback: (url) => {
          Browser.open({ url });
      },
    })
```
